### ISSUE TITLE
Missing type annotations needed diagnostic

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#22567]
This is a fun one

```rust
struct S {
    field: (),
}

fn f() {
    let c = |s| {
        s.field;
    };
    c(S { field: () });
}

```

rustc complains about the closure parameter needing a type annotation, we don't. In fact we do infer it to be `S` but we do not actually infer the `field` access correctly, the field name there is unresolved for us, presumably we fall back the closure parameter to S later on, where as rustc bails out early. Would be cool to recover such that we can report the diagnostic while keeping the inference we have right now.

**Comment by ChayimFriedman2** [comment]
It actually diagnoses fine for me; are you sure you have the diagnostic enabled?

<img width="1105" height="330" alt="Image" src="https://github.com/user-attachments/assets/6c5c43c1-e343-496f-8abe-7fe722dafceb" />

**Comment by Veykril** [comment]
ah, sorry, I didn't express myself correctly, we should show the diagnostic at the binding creation site like rustc i think, that is the closure param

**Comment by ChayimFriedman2** [comment]
rustc shows at both, but this is badly suited for IDE. And the real source of the error is the field access, since it requires the type to be known immediately.
